docs/spark-connect-extending.md

```diff
@@ -0,0 +1,248 @@
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+---
+
+Apache Spark provides different ways for users and developers to extend the
+system with custom functionality using many different extension points. Due to
+the separation of the client from the server, some of the extensions mechanism
+are changing.
+
+This can be separated into two main categories: First, extensions that are
+purely client side and do not extend the way the Spark server is behaving.
+Second, extensions that interact directly with Spark and need to run directly on
+the driver.
+
+
+## Client Extensions
+
+- Client extensions are programs that operate purely on the client surface of
+  Apache Spark and do not require any server side changes.
+
+
+## Server Extensions
+
+One of the biggest benefits of Spark Connect is the ability to extend Apache
+Spark with extensions and then use them seamlessly in all clients from any
+programming language with relative minimal effort.
+
+Building an extension for Spark Connect is an approach aimed at enhancing the
+functionality and flexibility of Apache Spark. Spark Connect operates on three
+core primitives: relations, expressions, and commands, each serving a unique
+purpose within the data processing framework.
+
+Relations in Spark Connect are fundamental to dataset transformations, acting as
+the mechanism through which an optional input dataset is transformed into an
+output dataset. Conceptually, relations can be likened to tables within a
+database, manipulated to achieve desired outcomes. Their functionality closely
+mirrors that of the DataFrame API, providing a familiar and intuitive interface
+for data manipulation.
+
+Expressions form another critical component of Spark Connect, functioning as
+operations that can be applied to individual columns or a set of columns within
+a dataset. These expressions are akin to functions in programming, offering a
+level of granularity in data processing that is comparable to the operations
+available through the Column API. This allows users to easily extend Spark with
+custom expression functionality and make it available to all clients.
+
+Commands stand out within Spark Connect as distinct actions that can be
+executed. Unlike relations, which focus on the transformation and nesting of
+output data, commands represent singular operations that perform specific tasks
+on the data. An example of such a command is a Data Manipulation Language (DML)
+command, which enables direct manipulation of the data stored within Spark. This
+differentiation highlights the versatility and power of Spark Connect, making it
+an essential tool for comprehensive data processing and analysis.
```

Comment:
I think you can drop the final sentence (from "This differentiation" to the end of the paragraph) since it repeats the message of the following paragraph.